From d2ee27aff05fedbba20bb3cafa532f11cccce89a Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:02:00 -0400
Subject: [infiniband] mlx4: Don't duplicate the default RoCE GID

Message-id: <c21cbbe17ea929e5b6e30c2b8807ba86e2ac1f26.1410425016.git.avadai@redhat.com>
Patchwork-id: 93691
O-Subject: [Patch RHEL7.1 V1 184/197] IB/mlx4: Don't duplicate the default RoCE GID
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Moni Shoua <monis@mellanox.com>

When reading the IPv6 addresses from the net-device, make sure to avoid
adding duplicate entry to the GID table as of equality between the default
GID we generate to the default IPv6 link-local address of the device.

Fixes: acc4fcc ('IB/mlx4: Make sure GID index 0 is always occupied')
Signed-off-by: Moni Shoua <monis@mellanox.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>

Upstream: f5c4834d9328c4ed9fe5dcbec6128d6da16db69a
Signed-off-by: Amir Vadai <avadai@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/mlx4/main.c b/drivers/infiniband/hw/mlx4/main.c
index 2421ce6..c962394 100644
--- a/drivers/infiniband/hw/mlx4/main.c
+++ b/drivers/infiniband/hw/mlx4/main.c
@@ -1676,6 +1676,7 @@ static void mlx4_ib_get_dev_addr(struct net_device *dev,
  struct inet6_dev *in6_dev;
  union ib_gid  *pgid;
  struct inet6_ifaddr *ifp;
+ union ib_gid default_gid;
 #endif
  union ib_gid gid;
 
@@ -1696,12 +1697,15 @@ static void mlx4_ib_get_dev_addr(struct net_device *dev,
   in_dev_put(in_dev);
  }
 #if IS_ENABLED(CONFIG_IPV6)
+ mlx4_make_default_gid(dev, &default_gid);
  /* IPv6 gids */
  in6_dev = in6_dev_get(dev);
  if (in6_dev) {
   read_lock_bh(&in6_dev->lock);
   list_for_each_entry(ifp, &in6_dev->addr_list, if_list) {
    pgid = (union ib_gid *)&ifp->addr;
+   if (!memcmp(pgid, &default_gid, sizeof(*pgid)))
+    continue;
    update_gid_table(ibdev, port, pgid, 0, 0);
   }
   read_unlock_bh(&in6_dev->lock);
-- 
1.7.1